### PR TITLE
Multi task mzr

### PR DESCRIPTION
**Pull request by marouaneamz**

### CODE REVIEW

File: mmcls/evaluation/metrics/multi_task.py [hunk 1]
```diff
@@ -101,38 +70,25 @@ def process(self, data_batch, data_samples: Sequence[dict]):
...
+                    filtered_data_samples.append(data_sample[task_name])
+            for metric in self._metrics[task_name]:
+                # Current implementation is only comptaible
+                # With 2 types of metrics :
+                # * Cls Metrics
+                # * Nested Cls Metrics
+                # In order to make it work with other
+                # non-cls heads/metrics, one will have to
+                # override the current implementation
+                metric.process(data_batch, filtered_data_samples)
```
Comment (piercus) [inline mmcls/evaluation/metrics/multi_task.py:87]: Please mask the data_batch here

File: tests/test_evaluation/test_metrics/test_multi_task_metrics.py [hunk 2]
```diff
@@ -4,63 +4,81 @@
...
+            else:
+                sample[task_name] = {}
+                for task_name2 in score[task_name]:
+                    task_sample = ClsDataSample().set_pred_score(
+                        score[task_name][task_name2])
+                    task_sample.set_gt_label(label[task_name][task_name2])
+                    sample[task_name][task_name2] = task_sample.to_dict()
+        pred2.append(sample)
 
+    print(pred2)
```
Comment (piercus) [inline tests/test_evaluation/test_metrics/test_multi_task_metrics.py]: remove

File: tests/test_structures/test_datasample.py [hunk 3]
```diff
@@ -122,137 +122,3 @@ def test_set_pred_score(self):
...
-
-    def _test_set_label(self, key):
-        data_sample = MultiTaskDataSample()
-        method = getattr(data_sample, 'set_' + key)
-        # Test Dict without metainfo
-        method({'task0': 0, 'task1': 2})
-        self.assertIn(key, data_sample)
-        label = getattr(data_sample, key)
-        self.assertIsInstance(label, LabelData)
-        self.assertEqual(getattr(label, 'task0'), 0)
```
Comment (piercus) [inline tests/test_structures/test_datasample.py]: Please create a little test at least to cover the MultiTaskDataSample init

**Review by piercus — CHANGES_REQUESTED** [review]

File: mmcls/models/heads/multi_task_head.py [hunk 4]
```diff
@@ -136,26 +114,25 @@ def predict(
...
-
-    def _get_predictions(self, preds_dict, data_samples):
-        """Post-process the output of MultiTaskHead."""
-        pred_dicts = [
-            dict(zip(preds_dict, t)) for t in zip(*preds_dict.values())
-        ]
+            if data_samples is None:
+                task_samples = None
+            else:
+                task_samples = [item.get(task_name) for item in data_samples]
```
Comment (marouaneamz) [inline mmcls/models/heads/multi_task_head.py:123]: if the task does not exist, it will return None , and it is impossible to do set_pred for Nones

**Review by piercus — APPROVED** [review]

**Comment by piercus** [comment]
FAILED tests/test_evaluation/test_metrics/test_multi_task_metrics.py::MultiTaskMetric::test_evaluate - TypeError: list indices must be integers or slices, not str

File: mmcls/models/heads/multi_task_head.py (hunk 4 above)
Comment (marouaneamz) [reply mmcls/models/heads/multi_task_head.py:123]:
```
File "tools/train.py", line 164, in <module>
    main()
  File "tools/train.py", line 160, in main
    runner.train()
  File "/home/marouane/.local/lib/python3.8/site-packages/mmengine/runner/runner.py", line 1659, in train
    model = self.train_loop.run()  # type: ignore
  File "/home/marouane/.local/lib/python3.8/site-packages/mmengine/runner/loops.py", line 94, in run
    self.runner.val_loop.run()
  File "/home/marouane/.local/lib/python3.8/site-packages/mmengine/runner/loops.py", line 343, in run
    self.run_iter(idx, data_batch)
  File "/home/marouane/.local/lib/python3.8/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/home/marouane/.local/lib/python3.8/site-packages/mmengine/runner/loops.py", line 363, in run_iter
    outputs = self.runner.model.val_step(data_batch)
  File "/home/marouane/.local/lib/python3.8/site-packages/mmengine/model/base_model/base_model.py", line 133, in val_step
    return self._run_forward(data, mode='predict')  # type: ignore
  File "/home/marouane/.local/lib/python3.8/site-packages/mmengine/model/base_model/base_model.py", line 301, in _run_forward
    results = self(**data, mode=mode)
  File "/home/marouane/.local/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/marouane/dev/mmclassification1.x/mmcls/models/classifiers/image.py", line 119, in forward
    return self.predict(inputs, data_samples)
  File "/home/marouane/dev/mmclassification1.x/mmcls/models/classifiers/image.py", line 244, in predict
    return self.head.predict(feats, data_samples, **kwargs)
  File "/home/marouane/dev/mmclassification1.x/mmcls/models/heads/multi_task_head.py", line 124, in predict
    task_samples = head.predict(feats, task_samples)
  File "/home/marouane/dev/mmclassification1.x/mmcls/models/heads/cls_head.py", line 130, in predict
    predictions = self._get_predictions(cls_score, data_samples)
  File "/home/marouane/dev/mmclassification1.x/mmcls/models/heads/cls_head.py", line 144, in _get_predictions
    data_sample.set_pred_score(score).set_pred_label(label)
AttributeError: 'NoneType' object has no attribute 'set_pred_score'

```

File: mmcls/models/heads/cls_head.py [hunk 5]
```diff
@@ -138,14 +138,16 @@ def _get_predictions(self, cls_score, data_samples):
...
+            data_samples = [None for _ in range(pred_scores.size(0))]
+
+        for data_sample, score, label in zip(data_samples, pred_scores,
+                                             pred_labels):
+            data_sample_ = data_sample
+            if data_sample is None:
+                data_sample_ = ClsDataSample()
+
+            data_sample_.set_pred_score(score).set_pred_label(label)
+            out_data_samples.append(data_sample_)
```
Comment (piercus) [inline mmcls/models/heads/cls_head.py]:
```suggestion
            if data_sample is None:
                data_sample = ClsDataSample()

            data_sample.set_pred_score(score).set_pred_label(label)
            out_data_samples.append(data_sample)
```

**Review by piercus — CHANGES_REQUESTED** [review]

**Pull request merged**